### ISSUE TITLE
Expand window bandwidth dictionary

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#399]
Right now, we only have Hann window in there, but we could add all of the scipy window functions.

Side note: the value currently recorded is half width half max. Would it be clearer to change this to be full width half max, and modify cqt accordingly? (Thanks to @lostanlen for pointing this out)

**Comment by lostanlen** [comment]
Actually, I'm starting to think that for energetic interpretation, the most appropriate would probably be equivalent rectangular bandwdith (called equivalent noise bandwidth in the Harris 1976 paper) instead of bandwidth at half-maximum.

**Comment by bmcfee** [comment]
That seems reasonable enough.  Would that lower the cutoff frequency for CQT compared to the current implementation?

**Comment by lostanlen** [comment]
I'm not sure of what you mean. The ERB is ~5% larger than the FWHM for most windows. So the CQT filters will be slightly more localized in the Fourier domain than the curent implementation.

**Comment by bmcfee** [comment]
The `window_bandwidth` function isn't used in basis construction; only to check that the resulting filters don't get too close to nyquist.  If that bandwidth is larger, then there's a higher chance of triggering the nyquist exception; this means that the effective upper bound on frequencies will be lower than if we use FWHM.  This would break backwards compatibility -- I'm okay with that, but I want to understand exactly how much of a change it would introduce.

My question is: how much difference is there here for Hann?

**Comment by lostanlen** [comment]
OK, thanks for clarifying. The bandwidth is 2.5% larger for Hann. So the cutoff frequency is 2.5% lower. That's about 43 cents of a semitone.

**Comment by lostanlen** [comment]
Regarding asymmetric triangular filters, which are used in mel-spectrograms, we could define them with a parameter r = the ratio of the right half-bandwidth to the left half-bandwidth.
What is nice is that the ERB of such a window is equal to 4/3 independently from r.
In the context of a constant-Q filter bank, r = 2^(1/Q)

**Comment by bmcfee** [comment]
Circling back on this after spending some time this morning implementing the calculation for generic filters.

I think I'm on board with switching over to using ENBW for this.  Aside from the energy interpretation that @lostanlen mentioned above, it has the added benefit of being much simpler to compute.  Computational complexity hasn't mattered much in the past for this, since we only had a single hard-coded value (hann).

However, if we add a function to compute ENBW from an arbitrary window, we can support custom window functions, or anything supported by `get_window()` including extended parametric windows (gaussian, kaiser, etc.).  I think this benefit outweighs the loss of half a semitone at the high end.

Of course, we could also implement the -3dB cutoff calculation and support arbitrary windows without modifying the current behavior.  The reason I'm (slightly) against this is that to do it correctly seems to require some much heavier machinery (fft + interpolation), and I suspect the resulting code would be slower, more brittle, and difficult to maintain.

What do y'all think? @lostanlen @ebattenberg @dpwe

**Comment by lostanlen** [comment]
Practically speaking, my use cases are:
* rectangular
* Hann
* triangulary asymmetric
So I don't need an arbitrary window to be computed off the shelf.

What is badly missing from the window collection from `scipy.signal` is the Fourier transform of a Gammatone wavelet (it's holomorphic BTW). Gammatones, as opposed to Morlets, cause slightly less pre-echo in some analysis-synthesis pipelines. They have also been used as the foundaion of cepstral coefficients (search GWCC). That said, @bmcfee has well pointed that introducing them would complexify the filterbank machinery. I agree with him.

**Comment by bmcfee** [comment]
> So I don't need an arbitrary window to be computed off the shelf.

Okay, but I think it would still be easier to support arbitrary window functions for this so that we don't have to hard-code a bunch of constants, or update the dictionary every time we think of a new window function to support.  Of course, we can pre-compute the ones for which it makes sense (hann, triangular, rectangular, all of the simple `scipy.signal` ones), but if you wanted to do something like a kaiser window with user-specified `alpha`, it would have to be computed anyway.

**Comment by bmcfee** [comment]
@lostanlen I think PRs #413 and #414 put enough abstraction in place that it would be easy for you to add asymmetric triangle and gammatone windows.

To do so, you'd need to do the following:

- Implement the functions in `librosa.filters`, say they're called `triangle_asym(Nx, skew, sym=True)` and `gammatone(...)`
- Add an index dictionary that maps the function name strings to the functions
- Add the window bandwidth entries to the dict
- Fill in the `# TODO` line in `get_window` to call locally indexed window functions before falling back on `scipy.signal.get_window`.

At that point, you'd be able to do `librosa.cqt(..., window='gammatone')` with no additional effort.

Feel free to make a new issue and PR about this at your convenience (probably not until after #414 is merged).

**Comment by bmcfee** [comment]
Implemented as #414.
